Add initial xcodebuild.vim implementation

gfontenot opened the pull request:
This is the first working version of this plugin. It allows you to
dynamically find the project or workspace, as well as dynamically
finding the first available scheme (looking only in the project).

It also has support for xcpretty, and dynamically adds it to the end of
the command if it's available.

gfontenot committed:
Add initial xcodebuild.vim implementation

This is the first working version of this plugin. It allows you to
dynamically find the project or workspace, as well as dynamically
finding the first available scheme (looking only in the project).

It also has support for xcpretty, and dynamically adds it to the end of
the command if it's available.

gabebw commented on bin/find_scheme.sh:
```diff
@@ -0,0 +1,5 @@
+#!/usr/bin/env sh
+
+xcodebuild -list -project $@ \
```
I think `$@` should be in quotes: `"$@"`, which correctly passes through arguments with spaces like `find_scheme "hello there" "hey"` rather than turning it into three arguments.

gfontenot replied:
yep, good call. I quoted it in the vim script, but forgot to do it here.

gabebw commented on plugin/xcodebuild.vim:
```diff
@@ -0,0 +1,49 @@
+endfunction
+
+function! s:run_command(cmd)
+  echom a:cmd
```
Hmm, is this valuable? Does the next `execute` line drop people into a shell anyway, thus wiping the messages?

gfontenot replied:
probably not. It was there for debugging and I just didn't remove it. Will do.

gfontenot committed:
Quote passed argument when finding schemes

gfontenot committed:
Remove logging info from run_command

This was used in debugging and was accidentally left in.

gabebw commented on bin/find_scheme.sh:
```diff
@@ -0,0 +1,5 @@
+#!/usr/bin/env sh
+
+xcodebuild -list -project "$@" \
+  | awk '/Schemes:/ { getline; print }' \
```
What does `getline` do? Is this line just printing lines that match `/Schemes:/`?

gabebw replied:
Oh, or is it printing the line _after_ a line that matches `/Schemes:/`?

gfontenot replied:
yeah, exactly.

gabebw replied:
:100:

gabebw commented on plugin/xcodebuild.vim:
```diff
@@ -0,0 +1,48 @@
+endfunction
+
+function! s:scheme()
+  let scheme = "silent ! 'sh " . s:plugin_path . "/bin/find_scheme.sh \"" . s:project_file() . "\""
```
Why do you need to run this explicitly with `sh`? I'd assume that the shebang runs it with sh so you'd only need:

``` vim
let scheme = "silent ! . " . s:plugin_path . "/bin/find_scheme.sh \"" . s:project_file() . "\""
```

Note the space after `.`!

Also, am I crazy, or is there no closing single quote in this line?

gabebw replied:
You could also replace the `.` with `source`, which I think is not POSIX-compliant but is more readable and works everywhere you care about like sh/bash/zsh.

gfontenot replied:
Good call

gfontenot committed:
Replace `sh` with `source`

This script has a shebang, so there's no real need for us to explicitly
run it with `sh`. Instead we can use `source`.

Also, there wasn't any closing single quote, and it wasn't needed
anyway, so I'm getting rid of it.

christoomey commented on plugin/xcodebuild.vim:
```diff
@@ -0,0 +1,48 @@
+endfunction
+
+function! s:build_target()
+  let xcworkspaceFile = globpath(expand('.'), '*.xcworkspace')
```
I think you can just use `glob('*.xcworkspace')` here. `glabpath` is for globbing accross a comma separated list of paths.

gfontenot replied:
Oh, nice. Good to know.

christoomey commented on plugin/xcodebuild.vim:
```diff
@@ -0,0 +1,48 @@
+endfunction
+
+function! s:xcpretty()
+  if executable('xcpretty')
```
Nice!

christoomey commented:
Welcome to the wild and wacky world of Vimscript!

gfontenot committed:
Use `glob` over `globpath`

`globpath` takes a comma separated list of paths, where `glob` just
looks in the pwd for the pattern.
